### ISSUE TITLE
`await` triggers `blocks_in_conditions`

### ISSUE THREAD

**Issue by xxchan** [rust-lang/rust-clippy#12016]
### Summary

```
warning: in a `match` scrutinee, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a `let`
   --> src/stream/src/executor/rearranged_chain.rs:264:13
    |
264 |             match select(&mut stop_rearrange_rx, upstream.next()).await {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `let res = await; match res`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blocks_in_conditions

warning: in a `match` scrutinee, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a `let`
   --> src/stream/src/executor/temporal_join.rs:296:13
    |
296 |             match combined.next().await {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `let res = await; match res`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blocks_in_conditions

warning: in a `match` scrutinee, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a `let`
   --> src/stream/src/executor/union.rs:110:9
    |
110 |         match active.next().await {
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `let res = await; match res`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blocks_in_conditions

```

### Lint Name

blocks_in_conditions

### Reproducer

I tried this code:

```rust
<code>
```

I saw this happen:

```
<output>
```

I expected to see this happen:

### Version

```text
rustc 1.77.0-nightly (e4c626dd9 2023-12-25)
binary: rustc
commit-hash: e4c626dd9a17a23270bf8e7158e59cf2b9c04840
commit-date: 2023-12-25
host: aarch64-apple-darwin
release: 1.77.0-nightly
LLVM version: 17.0.6
```

**Comment by xxchan** [comment]
cc  @J-ZhengLi  Is this possibile to be caused by https://github.com/rust-lang/rust-clippy/pull/11853?

**Comment by J-ZhengLi** [comment]
> cc @J-ZhengLi Is this possibile to be caused by #11853?

Sadly yes...

Hmmm, weird, I'm sure I was specifically targeting match expressions with `MatchSource::Normal` only, I'll see what went wrong. @rustbot claim

**Comment by J-ZhengLi** [comment]
@xxchan I can't reproduce the issue, would you mind provide a minimal reproduce code?

**Comment by BlackDex** [comment]
I have issue with this also when running clippy on the Vaultwarden code base.

It generates warning like this, which makes no sense at all.

```
warning: in a `match` scrutinee, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a `let`
   --> src/api/identity.rs:642:20
    |
642 |   struct ConnectData {
    |  ____________________^
643 | |     #[field(name = uncased("grant_type"))]
644 | |     #[field(name = uncased("granttype"))]
645 | |     grant_type: String, // refresh_token, password, client_credentials (API key)
...   |
690 | |     #[field(name = uncased("authrequest"))]
691 | |     auth_request: Option<String>,
    | |________________________________^ help: try: `let res = Option<String>; match res`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blocks_in_conditions

warning: in a `match` scrutinee, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a `let`
  --> src/api/notifications.rs:55:22
   |
55 |   struct WsAccessToken {
   |  ______________________^
56 | |     access_token: Option<String>,
   | |________________________________^ help: try: `let res = Option<String>; match res`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blocks_in_conditions

```

**Comment by BlackDex** [comment]
Since this is going to be released in the upcoming Rust 1.76, i have to allow this lint for now i think.

**Comment by lucab** [comment]
I think I'm also hitting this, the repository is not public but the source and warnings look like this:

```rust
  #[instrument(name = "Database::list_foo_bars", skip(self), err)]
  async fn list_foo_bars(&self) -> anyhow::Result<Vec<Uuid>> {
    let bars: Vec<Uuid> =
      sqlx::query("SELECT id FROM foo WHERE is_bar = true")
        .map::<_, Uuid>(|ref row: PgRow| row.get(0))
        .fetch_all(&self.pool)
        .await?;
    Ok(bars)
  }
```

That results in:
```
error: in a `match` scrutinee, avoid complex blocks or closures with blocks; instead, move the block or closure higher and bind it with a `let`
   --> db.rs:1:1
    |
1   |     async fn list_foo_bars(&self) -> anyhow::Result<Vec<Uuid>> {
    |  ___________________________________________________________________^
2   | |     let bars: Vec<Uuid> =
3   | |       sqlx::query("SELECT id FROM foo WHERE is_bar = true")
4   | |         .map::<_, Uuid>(|ref row: PgRow| row.get(0))
...   |
7   | |     Ok(bars)
8   | |   }
    | |___^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blocks_in_conditions

```
